common/prices.go

```diff
@@ -0,0 +1,385 @@
+package common
+
+import (
+	"encoding/gob"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"math"
+	"net/http"
+	"os"
+	"path/filepath"
+	"reflect"
+	"runtime"
+	"sort"
+	"strconv"
+	"sync"
+	"time"
+
+	"github.com/sirupsen/logrus"
+)
+
+var (
+	// Map of all historical prices. Date as "yyyy-mm-dd" to price in cents
+	historicalPrices map[string]float64 = make(map[string]float64)
+
+	// The latest price
+	latestPrice float64 = -1
+
+	// Latest price was fetched at
+	latestPriceAt time.Time
+
+	// Mutex to control both historical and latest price
+	pricesRwMutex sync.RWMutex
+
+	// Full path of the persistence file
+	pricesFileName string
+)
+
+func fetchAPIPrice(url string, resultPath []string) (float64, error) {
+	resp, err := http.Get(url)
+	if err != nil {
+		return -1, err
+	}
+	defer resp.Body.Close()
+
+	body, err := io.ReadAll(resp.Body)
+	if err != nil {
+		return -1, err
+	}
+
+	var priceJSON map[string]interface{}
+	json.Unmarshal(body, &priceJSON)
+
+	for i := 0; i < len(resultPath); i++ {
+		d, ok := priceJSON[resultPath[i]]
+		if !ok {
+			return -1, fmt.Errorf("API error: couldn't find '%s'", resultPath[i])
+		}
+
+		switch v := d.(type) {
+		case float64:
+			return v, nil
+		case string:
+			{
+				price, err := strconv.ParseFloat(v, 64)
+				return price, err
+			}
+
+		case map[string]interface{}:
+			priceJSON = v
+		}
+
+	}
+
+	return -1, errors.New("path didn't result in lookup")
+}
+
+func fetchCoinbasePrice() (float64, error) {
+	return fetchAPIPrice("https://api.coinbase.com/v2/exchange-rates?currency=ZEC", []string{"data", "rates", "USD"})
+
+}
+
+func fetchCoinCapPrice() (float64, error) {
+	return fetchAPIPrice("https://api.coincap.io/v2/rates/zcash", []string{"data", "rateUsd"})
+}
+
+func fetchBinancePrice() (float64, error) {
+	return fetchAPIPrice("https://api.binance.com/api/v3/avgPrice?symbol=ZECUSDC", []string{"price"})
+}
+
+func fetchHistoricalCoingeckoPrice(ts *time.Time) (float64, error) {
+	dt := ts.Format("02-01-2006") // dd-mm-yyyy
+	url := fmt.Sprintf("https://api.coingecko.com/api/v3/coins/zcash/history?date=%s?id=zcash", dt)
+
+	return fetchAPIPrice(url, []string{"market_data", "current_price", "usd"})
+}
+
+// Median gets the median number in a slice of numbers
+func median(inp []float64) (median float64) {
+
+	// Start by sorting a copy of the slice
+	sort.Float64s(inp)
+
+	// No math is needed if there are no numbers
+	// For even numbers we add the two middle numbers
+	// and divide by two using the mean function above
+	// For odd numbers we just use the middle number
+	l := len(inp)
+	if l == 0 {
+		return -1
+	} else if l == 2 {
+		return (inp[0] + inp[1]) / 2
+	} else if l%2 == 0 {
+		return (inp[l/2-1] + inp[l/2+1]) / 2
+	} else {
+		return inp[l/2]
+	}
+}
+
+// fetchPriceFromWebAPI will fetch prices from multiple places, discard outliers and return the
+// concensus price
+func fetchPriceFromWebAPI() (float64, error) {

```

Comment:
There are two ways I'd like to see this function improved:

1. It's possible for this function to return -1 without an error. This happens if all of the API endpoints fail. It turns out that some code higher up the call stack will catch this, but this is error-prone because newly added APIs might fail to make the same check. It would be clearer if this code returned an error in that case rather than a value of -1 without an error. The code in the higher-level API function should then just check for the error condition.
2. I would like to have the property that at least two providers must agree on the price (within 20%, say). This is currently true of this function as-sis, but it relies on there being at least three providers in the list. If one is deleted, then the code will return that provider's value without it being cross-checked by another one. I suggest changing the code below so that at least to providers must agree.

Reply:
The latest commit implements both of these suggestions (thanks for them!). I manually tested the second suggestion by deleting one of the 3 providers, and since the remaining two are close, the result is their average (which is shown in the third line):
```
{"app":"lightwalletd","level":"info","method":"CurrentPrice","msg":"Service","price":114.2,"provider":"github.com/zcash/lightwalletd/common.fetchBinancePrice","time":"2022-05-17T19:14:35-06:00"}
{"app":"lightwalletd","level":"info","method":"CurrentPrice","msg":"Service","price":114.67531021588819,"provider":"github.com/zcash/lightwalletd/common.fetchCoinCapPrice","time":"2022-05-17T19:14:35-06:00"}
{"app":"lightwalletd","level":"info","method":"CurrentPrice","msg":"Service","price":114.4376551079441,"time":"2022-05-17T19:14:35-06:00"}
```
Then I added 100 to one of the provider's result (still only two providers), and the result is as desired:
```
{"app":"lightwalletd","level":"info","method":"CurrentPrice","msg":"Service","price":114.2,"provider":"github.com/zcash/lightwalletd/common.fetchBinancePrice","time":"2022-05-17T19:15:39-06:00"}
{"app":"lightwalletd","level":"info","method":"CurrentPrice","msg":"Service","price":214.7104225164866,"provider":"github.com/zcash/lightwalletd/common.fetchCoinCapPrice","time":"2022-05-17T19:15:39-06:00"}
{"app":"lightwalletd","error":"Discarding price (114.20) because too far away from median (164.46","level":"error","method":"CurrentPrice","msg":"Service","time":"2022-05-17T19:15:39-06:00"}
{"app":"lightwalletd","error":"Discarding price (214.71) because too far away from median (164.46","level":"error","method":"CurrentPrice","msg":"Service","time":"2022-05-17T19:15:39-06:00"}
{"app":"lightwalletd","level":"error","msg":"Error getting prices from web APIs: insufficient price providers are available","time":"2022-05-17T19:15:39-06:00"}
```

Reply:
I'm new to the changes and all the implementation here, but if we're relying on multiple providers does that mean we're subject to denial of service if even one of those services goes down temporarily?

Reply:
I verified that our requests to the providers do timeout after about 10 seconds, but during that time, gRPC requests are still processed (other than, of course, requests to `coingecko` for historical data, but in that case, only `GetZECPrice` gRPCs are delayed, not other gRPCs, this server is multithreaded). So a non-responding provider would not cause DoS to our (wallet) clients.